### DISCUSSION TITLE
Apply same function to same agents but with different states

### DISCUSSION THREAD

**Discussion by MatBiocentis** [FLAMEGPU/FLAMEGPU2#1132] [Q&A] [answered]
Dear all, I need your help. 
I am using pyFlameGpu for a vary basic simulation:

- I have an agent with an integer variable, say "age". The agent has different states (s0, s1, s2, s3). The initial state is s0. 
- I have defined a non-conditional function called "aging" that increases by one the variable age. 
- In addition, I have a three conditional functions that should update the agent's states accordingly to their ages. For example, if  age == 1: s0 -> s1, if age > 3: s1 -> s2, etc. However, it looks like that the aging function is called  only untill the agents are in the initial state (s0) and as soon as they progress to s1 the aging function is not executed anymore. Is there a way to force the execution of the "aging" function regardless agents' state?

Thanks a lot for your help.

Best regards,
Teo

**Comment by Robadob** [comment]
> Is there a way to force the execution of the "aging" function regardless agents' state?

You will need to redefine the agent function for each agent state.

Agent's in different states are not stored together in memory, so they will always execute agent functions distinct from other states.

**Reply by Robadob** [reply]
I've created an issue to track the feature you want: https://github.com/FLAMEGPU/FLAMEGPU2/issues/1133

I can't make any promises about priority.

**Reply by MatBiocentis** [reply]
Thanks @Robadob , I nailed it and it works well. 
Is it possible to declare multiple states for an agent? For example, the agent "a" has both state "s0" and state "WT", agent "b" has state "s0" and state "GT", etc.

**Reply by Robadob** [reply]
I'm not too sure what your mean with your question.

Agent states are specific to an agent, there is no conflict if two different agents both have a state with the same name they will still be separate.

**Reply by MatBiocentis** [reply]
Sorry @Robadob , what I meant is to initialize multiple states for a single agent at the same time.
I'll describe a possible scenario:

- Assume that we have an agent "a" with an integer variable called "age". 
- At the beginning of the simulation "age = 0" and the agent is identified by the following states ={"young", "unmated"}.
- The age will increase over the course of the simulation.
- During the simulation, the agent might interact with other agents and they can mate
- Thus, based on age and possibly on mating behavior, there will be some transitions during the simulation that will allow the agent to go from {"young", "unmated"} to {"adult","unmated"} to {"adult", "mated"} and may return to {"adult", "unmated"}

As far as I know, in the current implementation of flamegpu, an agent can be represented with only one state at a time e.g. {young} or {unmated} but not both. Am I wrong?

For completeness:
The following works:
_cudaSimulation = pyflamegpu.CUDASimulation(model)_
.....
_cudaSimulation.setPopulationData(agentPopulation,"young")__

What I am looking for is something like:
_cudaSimulation.setPopulationData(agentPopulation,["young","unmated"])_

**Reply by Robadob** [reply, marked as answer]
> As far as I know, in the current implementation of flamegpu, an agent can be represented with only one state at a time e.g. {young} or {unmated} but not both. Am I wrong?

Yes, multiple states is not something that's been implemented (or even considered iirc). It may make sense for you to make one or both of them variables rather than states depending on your model.

**Reply by MatBiocentis** [reply]
Sorry to bother you, @Robadob . I see your point and it is a pity that is not possible to declare multiple states simoultaneously. I will consider the variables. 

But, meantime I tried the following: 
_cudaSimulation.setPopulationData(agentPopulation,'young')
cudaSimulation.setPopulationData(agentPopulation,'unmated')_

and they work. It looks like that "young" will be used as initial state. I am not sure if this is going to duplicate the population or not, I don't think so, and a possible explaination is that I am passing the same pointer "agentPopulation" . It looks like that setPopulationData is "appending" the second state ("unmated") to the initial state. Right?

Now, during the simulation, I can access and iterate over the population's members by:
_cudaSimulation.step()
out_population = pyflamegpu.AgentVector(agents)
cudaSimulation.getPopulationData(out_population ,"young")
for ag in out_population: [...]_
or 
_cudaSimulation.step()
out_population = pyflamegpu.AgentVector(agents)
cudaSimulation.getPopulationData(out_population ,"unmated")
for ag in out_population: [...]_

Is it possible to combine the two access? Retrieve the youngs that are also unmated? A solution that I can implement is to retrieve the IDs of the two sub-populations and do the intersection and iterate on the final list, but still it is unclear to me how to use the "states"  for conditional functions since they have been declared and there is a population with them:

The following works:
_af1 = agent.newRTCFunction("E2L_Transition_Function", E2L_Transition_Function)
af1.setInitialState("young")
af1.setEndState("adult")
af1.setRTCFunctionCondition(Young_2_Adult_Transition)_

This does not work:
_af1 = agent.newRTCFunction("E2L_Transition_Function", E2L_Transition_Function)
af1.setInitialState("young")
af1.setInitialState("unmated")
af1.setEndState("adult")
af1.setRTCFunctionCondition(Young_2_Adult_Transition)_

**Reply by Robadob** [reply]
```c
cudaSimulation.setPopulationData(agentPopulation,'young')
cudaSimulation.setPopulationData(agentPopulation,'unmated')
```
This will add the contents of agentPopulation twice, once in the state `young`, once in the state `unmated`. It takes a copy (data is moved to the GPU), no pointers are involved.

**Reply by MatBiocentis** [reply]
My illusions were shattered.  Thanks a lot; I will switch to variables.
